sklearn/decomposition/tests/test_factor_analysis.py

```diff
@@ -83,3 +86,41 @@ def test_factor_analysis():
         precision = fa.get_precision()
         assert_array_almost_equal(np.dot(cov, precision),
                                   np.eye(X.shape[1]), 12)
+
+
+@pytest.mark.parametrize("data_type, expected_type", [
+    (np.float32, np.float32),
+    (np.float64, np.float64),
+    (np.int32, np.float64),
+    (np.int64, np.float64)
+])
+def test_lda_dtype_match(data_type, expected_type):
+    rng = np.random.RandomState(0)
+    n_samples, n_features, n_components = 20, 5, 3
+    W = rng.randn(n_components, n_features)
+    h = rng.randn(n_samples, n_components)
+    noise = rng.gamma(1, size=n_features) * rng.randn(n_samples, n_features)
+    X = np.dot(h, W) + noise
+    for method in ['randomized', 'lapack']:
+        clf = FactorAnalysis(n_components=n_components, svd_method=method)
+        clf.fit(X.astype(data_type))
+        assert clf.components_.dtype == expected_type
```

Comment:
The dtype is tested only for components_, but should any numpy attribute preserve the dtype float32 ? If so, should the test be modified ?